ozhera-log/log-agent/src/main/java/com/xiaomi/mone/log/agent/channel/WildcardChannelServiceImpl.java

```diff
@@ -0,0 +1,397 @@
+                    wrapDataToSend(remainMsg, readResult, patternCode, ip, Instant.now().toEpochMilli());
+                }
+            }
+        }, 30, 30, TimeUnit.SECONDS);
+    }
+
+    private void wrapDataToSend(String lineMsg, AtomicReference<ReadResult> readResult, String patternCode, String localIp, long ct) {
+        String filePathName = readResult.get().getFilePathName();
+        LineMessage lineMessage = createLineMessage(lineMsg, readResult, filePathName, patternCode, localIp, ct);
+        updateChannelMemory(channelMemory, filePathName, getLogTypeEnum(), ct, readResult);
+
+        lineMessageList.add(lineMessage);
+
+        int batchSize = msgExporter.batchExportSize();
+        if (lineMessageList.size() > batchSize) {
+            List<LineMessage> subList = lineMessageList.subList(0, batchSize);
+            doExport(subList);
+        }
+    }
+
+
+    private void doExport(List<LineMessage> subList) {
+        try {
+            if (CollectionUtils.isEmpty(subList)) {
+                return;
+            }
+            //Current limiting processing
+            chain.doFilter();
+
+            long current = System.currentTimeMillis();
+            msgExporter.export(subList);
+            logCounts += subList.size();
+            lastSendTime = System.currentTimeMillis();
+            channelMemory.setCurrentTime(lastSendTime);
+
+            log.info("doExport channelId:{}, send {} message, cost:{}, total send:{}, instanceId:{},", channelDefine.getChannelId(), subList.size(), lastSendTime - current, logCounts, instanceId());
+        } catch (Exception e) {
+            log.error("doExport Exception", e);
+        } finally {
+            subList.clear();
+        }
+    }
+
+    private void startExportQueueDataThread() {
+        scheduledFuture = ExecutorUtil.scheduleAtFixedRate(() -> {
+            // If the mq message is not sent for more than 10 seconds, it will be sent asynchronously.
+            if (System.currentTimeMillis() - lastSendTime < 10 * 1000 || CollectionUtils.isEmpty(lineMessageList)) {
+                return;
+            }
+            synchronized (lock) {
+                this.doExport(lineMessageList);
+            }
+        }, 10, 7, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public ChannelDefine getChannelDefine() {
+        return channelDefine;
+    }
+
+    @Override
+    public ChannelMemory getChannelMemory() {
+        return channelMemory;
+    }
+
+    @Override
+    public Map<String, Long> getExpireFileMap() {
+        return Maps.newHashMap();
+    }
+
+    @Override
+    public void cancelFile(String file) {
+
+    }
+
+    @Override
+    public Long getLogCounts() {
+        return logCounts;
+    }
+
+    @Override
+    public void refresh(ChannelDefine channelDefine, MsgExporter msgExporter) {
+        this.channelDefine = channelDefine;
+        if (null != msgExporter) {
+            synchronized (this.lock) {
```

Comment:
It's recommended to use ReentrantLock instead of synchronized, otherwise it may cause pinning issues in coroutine scenarios.

Reply:
ok,I will update it based on ReentrantLock